tests/devices/test_default_mixed.py

```diff
@@ -987,6 +987,57 @@ def test_apply_pauli_error(self, tol):
 
         assert np.allclose(dev.state, target, atol=tol, rtol=0)
 
+@pytest.mark.parametrize("nr_wires", [2,3])
+class TestReadoutError:
+    """Tests for measurement readout error"""
+
+    def test_readout_prob_0(self,nr_wires):
+        """Tests the measurement results for readout error probability=0"""
+        dev = qml.device("default.mixed",wires=nr_wires,readout_prob=0)
+
+        @qml.qnode(dev)
+        def circuit1():
+            return qml.expval(qml.PauliZ(0)), qml.expval(qml.PauliZ(1))
+
+        res=circuit1()
+        expected = np.array([1, 1])
+        assert np.allclose(res,expected)
+
+    def test_readout_prob_1(self,nr_wires):
+        """Tests the measurement results when readout error probability=1"""
+        dev = qml.device("default.mixed",wires=nr_wires,readout_prob=1)
+
+        @qml.qnode(dev)
+        def circuit2():
+            return qml.expval(qml.PauliZ(0)), qml.expval(qml.PauliZ(1))
+
+        res=circuit2()
+        expected = np.array([-1,-1])
+        assert np.allclose(res,expected)
+
+    def test_readout_prob_half(self,nr_wires):
+        """Tests the measurement results on PauliZ operator when readout error probability=0.5"""
+        dev = qml.device("default.mixed",wires=nr_wires,shots=10000,readout_prob=0.5)
+
+        @qml.qnode(dev)
+        def circuit3():
+            return qml.expval(qml.PauliZ(0))
+
+        res=circuit3()
+        # setting the tolerance in expectation value to be 5 per cent.
+        tol1 = 0.05
+        expected = 0
+        assert abs(res-expected) < tol1
```

Comment:
Could this test be stochastic? I.e., could it be that if we run it 100 times, it will fail at least once?

If so, we use the `flaky` package that helps with such tests. We can create rules such as: run the test such that we require that from a max. of 5 runs we would like it to pass at least 3 times. If that doesn't happen, then something went wrong.

Reply:
Thanks for the comment. That makes sense to use some package like `flaky` when the expected result is not always exactly the same in different runs.

Nonetheless, I had another thought regarding the test case `test_readout_prob_half `. When we provide a value for `shots`, then the expected result has a slight variance in different runs. I think I can just remove `shots=10000` from `dev` in  `test_readout_prob_half ` method. Then the computation will be exact as in the other test cases I wrote for readout_prob=0 and 1. In that case, I can use `np.allclose` in all those 3 test cases. Do you think writing tests only with exact computations (without providing `shots`)  is sufficient for the readout error or do we need test cases with a `shots` input as well?